proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+* Review: ([pitch](https://forums.swift.org/t/pitch-isolated-default-value-expressions/67714))
+
+## Introduction
+
+Default value expressions are permitted for default arguments and default stored property values. There are several issues with the current actor isolation rules for default value expressions: the rules for stored properties admit data races, the rules for default argument values are overly restrictive, and the rules between the different places you can use default value expressions are inconsistent with each other, making the actor isolation model harder to understand. This proposal unifies the actor isolation rules for default value expressions, eliminates data races, and improves expressivity by safely allowing isolation for default values.
+
+## Motivation
+
+The current actor isolation rules for initial values of stored properties admit data races. For example, the following code is currently valid:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+class C {
+  @MainActor var x1 = requiresMainActor()
+  @AnotherActor var x2 = requiresAnotherActor()
+
+  nonisolated init() {} // okay???
+}
+```
+
+The above code allows any context to initialize an instance of `C()` through a synchronous, non-isolated `init` that synchronously calls both a `@MainActor`-isolated and a `@AnotherActor`-isolated function, violating actor isolation checking and enabling `requiresMainActor()` and `requiresAnotherActor()` to run concurrently with other code on those respective actors.
+
+Similarly, the current actor isolation rules for default argument values do not admit data races, but default argument values are always `nonisolated` which is overly restrictive. This rule prohibits programmers from making `@MainActor`-isolated calls in default argument values of `@MainActor`-isolated functions that are only ever called from the main actor. For example, the following code is not valid even though it is perfectly safe:
+
+```swift
+@MainActor class C { ... }
+
+@MainActor func f(c: C = C()) { ... } // error
+
+@MainActor func useFromMainActor() {
+  f()
+}
+```
+
+## Proposed solution
+
+I propose allowing default value expressions to require the caller to meet an isolation requirement in order to use the default value. The isolation requirement is inferred from the default value expression. If the caller does not meet the isolation requirement, then a value must be written explicitly for the argument or the stored property. This rule makes the stored property example above invalid at the point of the `nonisolated` initializer, because the isolation requirement of the default values for the stored properties is not satisfied. This rule also makes the default argument example above valid, because the `@MainActor` isolation requirement for the default argument of `f` is satisfied by the caller.
+
+## Detailed design
```

Comment:
Given the discussion on the pitch, it's probably worth adding a discussion of the order of evaluation for call arguments, because it's strongly influenced the thinking of this proposal.

Reply:
I've added a section on argument evaluation.